### ISSUE TITLE
Thinking about disaster recovery

### ISSUE THREAD

**Issue by bingkunyangvungle** [Aiven-Open/tiered-storage-for-apache-kafka#534]
# What is currently missing?

Since we have all the data needed in S3 already (including the [log data and the metadata](https://github.com/Aiven-Open/tiered-storage-for-apache-kafka/issues/158)), it can be of great help if we can just recover from the shutdown of the cluster, and launch the new cluster to use the same topic folder(prefix_randomstring) in S3. For the new cluster, we can consume from the oldest offset from the old topic folder. In this way, we can fully achieve the disaster recovery.

# How could this be improved?

Some suggestions:
1.Enable the topic can be created with designated folder (with random string) in S3 in the new cluster;
2.The new message for the topic can be stored in the older folder

# Is this a feature you would work on yourself?

I haven't take a look for the actual code part, but if someone can point me to the proper code link, I'd be happy to start looking.

**Comment by ivanyu** [comment]
Hi @bingkunyangvungle.
Such restoration is certainly possible. Do you see particular changes need to be done to this RSM plugin?

**Comment by ivanyu** [comment]
Kafka doesn't really support this out of the box and quite a bit of work, sometimes hacky, needs to be done on a fresh cluster to make it correctly pick up an "attached" remote storage. Start from that topic IDs must match between the remote storage and the cluster, but the user doesn't control IDs of topics being created.

**Comment by bingkunyangvungle** [comment]
This is the tricky part that I think it might be. One proposal that I can think of is to have an internal mapping between the `kafka-created topic ID <------> the plugin-managed topic ID`. Then the plugin will manage the remote storage folder with the plugin-managed topic ID. 
Also user can provide/configure the plugin-managed topic ID for topic provisioning. Of course the plugin would also need to check if the remote storage folder exist or not. 

Just toss some idea here to share.

**Comment by funky-eyes** [comment]
> Kafka doesn't really support this out of the box and quite a bit of work, sometimes hacky, needs to be done on a fresh cluster to make it correctly pick up an "attached" remote storage. Start from that topic IDs must match between the remote storage and the cluster, but the user doesn't control IDs of topics being created.

Perhaps we can ignore the topicid in` ObjectKeyFactory#mainPath`? Adding some configuration to achieve this purpose.

**Comment by ivanyu** [comment]
This probably also doesn't have to do to the plugin. The plugin is a pretty passive component and does only a limited number of operations by the broker request. I think this should be a separate tool and/or certain broker modifications to support restoration like this in the first place.

We're exploring this idea in Aiven, but for the reason mentioned above, I don't think the plugin will undergo much change in the course of this.

**Comment by jrmcclurg** [comment]
I recently started thinking about this type of disaster recovery after experimenting with deleting persistent disks in my Azure test cluster. I would like to at least come up with a script(s) that can re-create topic data locally from a specified topic(s) in blob storage.

Has functionality like this already been implemented? If not, do you see any issues with the following general approach?

1. Construct a `RemoteLogSegmentMetadata` from the `.rsm-manifest` file in remote storage.
2. Use `RemoteStorageManager.fetchLogSegment(...)` to download the `.log` file, and similarly use `fetchIndex(...)` and `fetchManifest(...)` to get the `.index` and `.manifest` files respectively.
3. Not sure how to reconstruct the `.timeindex` files.

**Comment by jrmcclurg** [comment]
@giuseppelillo Many thanks, this is extremely helpful! On a slightly-related note, would I also need to reconstruct the `__remote_log_metadata` (and also `__cluster_metadata`) topics based on the remote segments, or is Kafka able to recover without these?

**Comment by giuseppelillo** [comment]
@jrmcclurg yes you absolutely need to reconstruct the `__remote_log_metadata` topic.

I'm not sure about `__cluster_metadata` instead, I guess it depends on the order of operations. For example let's say you have some remote segment stored in object storage for partition `topic-0`. If you want that this topic to be present (and to contain all the data stored on object storage) as soon as you start your disaster recovery Kafka cluster, then yes you will also have to reconstruct `__cluster_metadata` so that the brokers immediately know about `topic-0`.
